src/main/java/shef/servlets/UserServlet.java

```diff
@@ -92,36 +94,49 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
+      
+    // Get properties that don't come from the request.
     String id = userService.getCurrentUser().getUserId();
     String email = userService.getCurrentUser().getEmail();
-    String username = request.getParameter("username-input");
-    String location = request.getParameter("location-input");
-    String profilePictureUrl = getUploadedFileUrl(request, "profile-pic-upload");
-    String bio = request.getParameter("bio-input");
+    String profilePicUrl = getUploadedFileUrl(request, "profile-pic-url");
+
+    // Get properties from the request.
+    Map<String, String[]> parameterMap = request.getParameterMap();
+    Set entrySet = parameterMap.entrySet();
+    Iterator it = entrySet.iterator();
 
     Key userKey = KeyFactory.createKey("User", id);
+    String keyString = KeyFactory.keyToString(userKey);
+    Entity user;
+    String redirectUrl;
 
-    if (profilePictureUrl == null) {
-      try {
-        Entity originalEntity = datastore.get(userKey);
-        profilePictureUrl = (String) originalEntity.getProperty("profile-picture-url");
-      } catch (EntityNotFoundException e) {
-        throw new IOException("No profile picture provided.");
-      }
+    try {
+      // Get existing user with the key.
+      user = datastore.get(userKey);
+      redirectUrl = "/profile-page.html?key=" + keyString;
+    } catch (EntityNotFoundException e) {
+      // Create a new User entity with data from the request.
+      user = new Entity(userKey);
+      redirectUrl = "/account-creation-finish.html";
     }
 
-    // Create a new User entity with data from the request.
-    Entity userEntity = new Entity(userKey);
-    userEntity.setProperty("email", email);
-    userEntity.setProperty("username", username);
-    userEntity.setProperty("location", location);
-    userEntity.setProperty("profile-picture-url", profilePictureUrl);
-    userEntity.setProperty("bio", bio);
-
-    // Store the User entity in Datastore.
-    datastore.put(userEntity);
-    response.sendRedirect("/user-list-test.html");
+    // These properties don't come from the request, so we add them here.
+    user.setProperty("email", email);
+    if(profilePicUrl != null) {
+      user.setProperty("profile-pic-url", profilePicUrl);
+    }
+
+    Entity finalUser = user.clone();
```

Comment:
Why do you need to clone?

Reply:
I did that because if I try to just use the original user object, I get this error when compiling: "local variables referenced from a lambda expression must be final or effectively final".

Reply:
The problem comes from the fact that I need to seperately add the email and profilepictureUrl to the user Entity, because they don't come from the request params. So me editing the user Entity makes it no longer "effectively final".

Reply:
Disregard what I said. I was able to remove the clone.